hilti/toolchain/src/compiler/codegen/codegen.cc

```diff
@@ -40,74 +40,149 @@ struct GlobalsVisitor : hilti::visitor::PreOrder<void, GlobalsVisitor> {
-                .result = "auto",
-                .id = {ns, "__globals"},
-                .args = {},
-                .linkage = "static",
-                .inline_body = body,
-            };
+        auto ns = cxx::ID(cg->options().cxx_namespace_intern, module_id);
 
-            unit->add(body_decl);
+        for ( const auto& g : globals ) {
+            auto cxx_g = g;
+            cxx_g.id = cxx::ID{ns, g.id.local()};
+            cxx_g.type = fmt("std::unique_ptr<%s>", g.type);
+            cxx_g.init = {};
+            cxx_g.linkage = "extern";
+            unit->add(cxx_g);
         }
+    }
 
-        if ( ! v.globals.empty() && include_implementation ) {
-            // Create the initGlobals() function.
-            auto id = cxx::ID{ns, "__init_globals"};
+    // Helper creating function initializing the module's globals.
+    void createInitGlobals(const Node& module, const ID& module_id, cxx::Unit* unit) {
+        auto ns = cxx::ID(cg->options().cxx_namespace_intern, module_id);
+        auto id = cxx::ID{ns, "__init_globals"};
 
-            auto body_decl =
-                cxx::declaration::Function{.result = cg->compile(type::void_, codegen::TypeUsage::FunctionResult),
-                                           .id = id,
-                                           .args = {{.id = "ctx", .type = "::hilti::rt::Context*"}},
-                                           .linkage = "extern"};
+        auto body_decl =
+            cxx::declaration::Function{.result = cg->compile(type::void_, codegen::TypeUsage::FunctionResult),
+                                       .id = id,
+                                       .args = {{.id = "ctx", .type = "::hilti::rt::Context*"}},
+                                       .linkage = "extern"};
 
-            auto body = cxx::Block();
-            cg->pushCxxBlock(&body);
+        auto body = cxx::Block();
+        cg->pushCxxBlock(&body);
 
-            if ( ! v.globals.empty() )
-                body.addStatement("::hilti::rt::detail::initModuleGlobals<__globals_t>(__globals_index)");
+        if ( cg->options().cxx_enable_dynamic_globals ) {
+            body.addStatement("::hilti::rt::detail::initModuleGlobals<__globals_t>(__globals_index)");
 
-            for ( auto g : v.globals ) {
+            for ( auto g : globals ) {
                 if ( g.init )
                     body.addStatement(fmt("__globals()->%s = %s", g.id.local(), *g.init));
                 else if ( g.args.size() )
                     body.addStatement(fmt("__globals()->%s = {%s}", g.id.local(), util::join(g.args, ", ")));
             }
+        }
+        else {
+            for ( const auto& g : globals ) {
+                auto cxx_g = g;
+                cxx_g.type = fmt("std::unique_ptr<%s>", g.type);
+                cxx_g.init =
+                    fmt("std::make_unique<%s>()",
+                        g.type); // initialize to a default value at first to match semantics of dynamic globals
```

Comment:
Naively I would have expected the naked values here. Could you add a comment why we need dynamic allocation instead? All runtime types should be default-constructible and to catch uninitialized access I'd image that a `std::optional` on the stack might perform better.

Reply:
I had that at first, but we can't control the time of destruction then, which needs to happen in `rt::done()` to be well- defined and match the existing semantics. Will add a comment.

Reply:
I see, if the reason was deterministic cleanup couldn't we with a `std::optional` instead to avoid the heap allocation and access? An optional can be `reset` which might be enough for our cleanup needs.

Reply:
I just changed it to `optional`, that's a good thought.